From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Tehin <62246450+Tehin@users.noreply.github.com>
Date: Wed, 2 Apr 2025 22:23:25 -0500
Subject: [PATCH] Minemora Patch


diff --git a/ca/spottedleaf/moonrise/patches/chunk_system/scheduling/ChunkHolderManager.java b/ca/spottedleaf/moonrise/patches/chunk_system/scheduling/ChunkHolderManager.java
index e03faa94df6c77ab4f3f5d9ffe107220387a0dde..84689078d096a02f306911a9863599830472de49 100644
--- a/ca/spottedleaf/moonrise/patches/chunk_system/scheduling/ChunkHolderManager.java
+++ b/ca/spottedleaf/moonrise/patches/chunk_system/scheduling/ChunkHolderManager.java
@@ -70,6 +70,9 @@ public final class ChunkHolderManager {
     public static final int ENTITY_TICKING_TICKET_LEVEL = ChunkLevel.ENTITY_TICKING_LEVEL;
     public static final int MAX_TICKET_LEVEL = ChunkLevel.MAX_LEVEL; // inclusive
 
+    // MoraPaper - Limit chunk unloads per tick
+    private static final int MAX_CHUNK_UNLOADS_PER_TICK = 180;
+
     public static final TicketType UNLOAD_COOLDOWN = ChunkSystemTicketType.create("chunk_system:unload_cooldown", null, 5L * 20L);
 
     private static final long NO_TIMEOUT_MARKER = Long.MIN_VALUE;
@@ -1176,7 +1179,9 @@ public final class ChunkHolderManager {
         // We do need to process updates here so that any addTicket that is synchronised before this call does not go missed.
         this.processTicketUpdates();
 
-        final int toUnloadCount = Math.max(50, (int)(unloadCountTentative * 0.05));
+        // MoraPaper - Limit chunk unload to prevent RAM overuse
+        final int toUnloadCount = Math.max(50,Math.min(MAX_CHUNK_UNLOADS_PER_TICK, (int) (unloadCountTentative * 0.05)));
+
         int processedCount = 0;
 
         for (final ChunkUnloadQueue.SectionToUnload sectionRef : unloadSectionsForRegion) {
diff --git a/net/minecraft/network/chat/OutgoingChatMessage.java b/net/minecraft/network/chat/OutgoingChatMessage.java
index 4635182c73f28f63ba84982cb6bcb422221b738e..eb40e9b8051c7e159146104383a747d1c3c26566 100644
--- a/net/minecraft/network/chat/OutgoingChatMessage.java
+++ b/net/minecraft/network/chat/OutgoingChatMessage.java
@@ -26,6 +26,9 @@ public interface OutgoingChatMessage {
             this.sendToPlayer(player, filtered, boundType, null);
         }
         public void sendToPlayer(ServerPlayer player, boolean filtered, ChatType.Bound boundType, @javax.annotation.Nullable Component unsigned) {
+            // MoraPaper
+            if (player.chatDisabled) return;
+
             player.connection.sendDisguisedChatMessage(unsigned != null ? unsigned : this.content, boundType);
             // Paper end
         }
@@ -43,6 +46,9 @@ public interface OutgoingChatMessage {
             this.sendToPlayer(player, filtered, boundType, null);
         }
         public void sendToPlayer(ServerPlayer player, boolean filtered, ChatType.Bound boundType, @javax.annotation.Nullable Component unsigned) {
+            // MoraPaper
+            if (player.chatDisabled) return;
+
             // Paper end
             PlayerChatMessage playerChatMessage = this.message.filter(filtered);
             playerChatMessage = unsigned != null ? playerChatMessage.withUnsignedContent(unsigned) : playerChatMessage; // Paper
diff --git a/net/minecraft/network/protocol/game/ClientboundPlayerInfoUpdatePacket.java b/net/minecraft/network/protocol/game/ClientboundPlayerInfoUpdatePacket.java
index 21d7cf384af69645b8ad61c74af4128f30811946..734d4b0da6d3dae04a56351b436aaff56452f349 100644
--- a/net/minecraft/network/protocol/game/ClientboundPlayerInfoUpdatePacket.java
+++ b/net/minecraft/network/protocol/game/ClientboundPlayerInfoUpdatePacket.java
@@ -229,7 +229,7 @@ public class ClientboundPlayerInfoUpdatePacket implements Packet<ClientGamePacke
                 player.getUUID(),
                 player.getGameProfile(),
                 listed, // Paper - Add Listing API for Player
-                player.connection.latency(),
+                player.connection == null ? player.fakePing : player.connection.latency(),
                 player.gameMode(),
                 player.getTabListDisplayName(),
                 player.isModelPartShown(PlayerModelPart.HAT),
diff --git a/net/minecraft/server/MinecraftServer.java b/net/minecraft/server/MinecraftServer.java
index 8b6541de8c891ee66cdfebd9f29dfc6eee7085d5..627ac5e745fac4adbb560d1ea0d1ac5af10431f2 100644
--- a/net/minecraft/server/MinecraftServer.java
+++ b/net/minecraft/server/MinecraftServer.java
@@ -1700,14 +1700,14 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
         this.getPlayerList().getPlayers().forEach(serverPlayer1 -> serverPlayer1.connection.suspendFlushing());
         this.server.getScheduler().mainThreadHeartbeat(); // CraftBukkit
         // Paper start - optimise Folia entity scheduler
-        ((io.papermc.paper.threadedregions.scheduler.FoliaGlobalRegionScheduler)org.bukkit.Bukkit.getGlobalRegionScheduler()).tick();
-        for (io.papermc.paper.threadedregions.EntityScheduler scheduler : this.entitySchedulerTickList.getAllSchedulers()) {
-            if (scheduler.isRetired()) {
-                continue;
-            }
-
-            scheduler.executeTick();
-        }
+        // ((io.papermc.paper.threadedregions.scheduler.FoliaGlobalRegionScheduler)org.bukkit.Bukkit.getGlobalRegionScheduler()).tick();
+        // for (io.papermc.paper.threadedregions.EntityScheduler scheduler : this.entitySchedulerTickList.getAllSchedulers()) {
+        //     if (scheduler.isRetired()) {
+        //         continue;
+        //     }
+
+        //     scheduler.executeTick();
+        // }
         // Paper end - optimise Folia entity scheduler
         io.papermc.paper.adventure.providers.ClickCallbackProviderImpl.ADVENTURE_CLICK_MANAGER.handleQueue(this.tickCount); // Paper
         io.papermc.paper.adventure.providers.ClickCallbackProviderImpl.DIALOG_CLICK_MANAGER.handleQueue(this.tickCount); // Paper
diff --git a/net/minecraft/server/level/ChunkMap.java b/net/minecraft/server/level/ChunkMap.java
index b7c16eeee03ac1183d3d26bd8ee1b13afd620f61..a01b60ce716456ae9ed1decf9d47b413b89c9b84 100644
--- a/net/minecraft/server/level/ChunkMap.java
+++ b/net/minecraft/server/level/ChunkMap.java
@@ -1213,11 +1213,16 @@ public class ChunkMap extends ChunkStorage implements ChunkHolder.PlayerProvider
             }
 
             if (lastChunkUpdate != currChunkUpdate || lastTrackedChunk != chunk) {
+                // MoraPaper - Iterate in the set itself not by creating a completely new arraylist
+                final java.util.Iterator<net.minecraft.server.network.ServerPlayerConnection> seenByIterator = this.seenBy.iterator();
+
                 // need to purge any players possible not in the chunk list
-                for (final ServerPlayerConnection conn : new java.util.ArrayList<>(this.seenBy)) {
+                while (seenByIterator.hasNext()) {
+                    final ServerPlayerConnection conn = seenByIterator.next();
                     final ServerPlayer player = conn.getPlayer();
                     if (!players.contains(player)) {
-                        this.removePlayer(player);
+                        seenByIterator.remove();
+                        this.forceRemovePlayerPairing(player);
                     }
                 }
             }
@@ -1321,6 +1326,12 @@ public class ChunkMap extends ChunkStorage implements ChunkHolder.PlayerProvider
             }
         }
 
+        // MoraSpigot - Remove the player without removing or checking for the seenBy, to prevent CME
+        // when the player is removed while in a loop
+        public void forceRemovePlayerPairing(ServerPlayer player) {
+            this.serverEntity.removePairing(player);
+        }
+
         public void updatePlayer(ServerPlayer player) {
             org.spigotmc.AsyncCatcher.catchOp("player tracker update"); // Spigot
             if (player != this.entity) {
diff --git a/net/minecraft/server/level/ServerChunkCache.java b/net/minecraft/server/level/ServerChunkCache.java
index 413f0673557b9e3f9177d15e9bef61bded209e34..7fcc82eaaa0b6dc87e5da4fbb16f7dd74c3fdffe 100644
--- a/net/minecraft/server/level/ServerChunkCache.java
+++ b/net/minecraft/server/level/ServerChunkCache.java
@@ -533,7 +533,23 @@ public class ServerChunkCache extends ChunkSource implements ca.spottedleaf.moon
         profiler.pop();
     }
 
+    // MoraPaper - Only do chunk mob spawning each 5 ticks instead of every tick
+    private final static int SPAWN_UPDATE_INTERVAL = 5;
+
+    // MoraPaper - The last time we ticked mob spawning
+    private int ticksSinceLastSpawnUpdate = 0;
+
     private void tickChunks(ProfilerFiller profiler, long timeInhabited) {
+        ticksSinceLastSpawnUpdate++;
+
+        // MoraPaper - Tick without the mob spawning between the interval
+        if (ticksSinceLastSpawnUpdate <= SPAWN_UPDATE_INTERVAL) {
+            iterateTickingChunksFaster();
+            return;
+        }
+
+        ticksSinceLastSpawnUpdate = 0;
+
         profiler.push("naturalSpawnCount");
         int naturalSpawnChunkCount = this.distanceManager.getNaturalSpawnChunkCount();
         // Paper start - Optional per player mob spawns
@@ -620,6 +636,15 @@ public class ServerChunkCache extends ChunkSource implements ca.spottedleaf.moon
         }
     }
 
+    // MoraPaper Start - Limit natural spawn radius
+    public static boolean isInNoNaturalSpawnRadius(ChunkPos pos) {
+        int radius = 10; //TODO from config
+        int x = pos.x;
+        int z = pos.z;
+        return x >= -radius && x < radius && z >= -radius && z < radius;
+    }
+    // MoraPaper End
+
     private void getFullChunk(long chunkPos, Consumer<LevelChunk> fullChunkGetter) {
         // Paper start - rewrite chunk system
         // note: bypass currentlyLoaded from getChunkNow
diff --git a/net/minecraft/server/level/ServerEntity.java b/net/minecraft/server/level/ServerEntity.java
index 638432a2e6506d3db6a25c068a33eeafb13cf0d6..ea19d987ef672b83c9dbe4dc39a0738f3cf01342 100644
--- a/net/minecraft/server/level/ServerEntity.java
+++ b/net/minecraft/server/level/ServerEntity.java
@@ -417,4 +417,5 @@ public class ServerEntity {
 
         void sendToTrackingPlayersFiltered(Packet<? super ClientGamePacketListener> packet, Predicate<ServerPlayer> filter);
     }
+
 }
diff --git a/net/minecraft/server/level/ServerLevel.java b/net/minecraft/server/level/ServerLevel.java
index 6c0b1ce3888ab599a84fb4a8ee69a0d6454e50fd..5b17919605874c33fc6ec6c2cf9f39b14bbbb7e9 100644
--- a/net/minecraft/server/level/ServerLevel.java
+++ b/net/minecraft/server/level/ServerLevel.java
@@ -184,7 +184,7 @@ public class ServerLevel extends Level implements ServerEntityGetter, WorldGenLe
     private static final Logger LOGGER = LogUtils.getLogger();
     private static final int EMPTY_TIME_NO_TICK = 300;
     private static final int MAX_SCHEDULED_TICKS_PER_TICK = 65536;
-    final List<ServerPlayer> players = Lists.newArrayList();
+    final List<ServerPlayer> players = new java.util.concurrent.CopyOnWriteArrayList<>(); // MoraPaper - Use CopyOnWriteArrayList for safe async loops
     public final ServerChunkCache chunkSource;
     private final MinecraftServer server;
     public final net.minecraft.world.level.storage.PrimaryLevelData serverLevelData; // CraftBukkit - type
@@ -585,6 +585,11 @@ public class ServerLevel extends Level implements ServerEntityGetter, WorldGenLe
     }
     // Paper end - chunk tick iteration
 
+    // MoraPaper - Get async entities from our tick list (calculated in another thread)
+    public java.util.concurrent.ConcurrentLinkedQueue<Entity> getEntitiesAsync() {
+        return entityTickList.getEntitiesAsync();
+    }
+
     public ServerLevel(
         MinecraftServer server,
         Executor dispatcher,
@@ -692,6 +697,8 @@ public class ServerLevel extends Level implements ServerEntityGetter, WorldGenLe
         this.chunkDataController = new ca.spottedleaf.moonrise.patches.chunk_system.io.datacontroller.ChunkDataController((ServerLevel)(Object)this, this.chunkTaskScheduler);
         // Paper end - rewrite chunk system
         this.getCraftServer().addWorld(this.getWorld()); // CraftBukkit
+        // MoraPaper
+        net.minemora.MoraPaper.startNearestPlayerUpdater(this);
     }
 
     // Paper start
@@ -1466,6 +1473,11 @@ public class ServerLevel extends Level implements ServerEntityGetter, WorldGenLe
         });
     }
 
+    public void disableDragon() {
+        this.dragonFight = null;
+        this.paperConfig().entities.spawning.scanForLegacyEnderDragon = false;
+    }
+
     public List<? extends EnderDragon> getDragons() {
         return this.getEntities(EntityType.ENDER_DRAGON, LivingEntity::isAlive);
     }
diff --git a/net/minecraft/server/level/ServerPlayer.java b/net/minecraft/server/level/ServerPlayer.java
index 00c77be8c02e5603bbb4207804d712a77994658b..6428095bae393be73f76686248e7753e8da3f845 100644
--- a/net/minecraft/server/level/ServerPlayer.java
+++ b/net/minecraft/server/level/ServerPlayer.java
@@ -230,6 +230,7 @@ public class ServerPlayer extends Player implements ca.spottedleaf.moonrise.patc
     public final ServerPlayerGameMode gameMode;
     private final PlayerAdvancements advancements;
     private final ServerStatsCounter stats;
+    public int fakePing;
     private float lastRecordedHealthAndAbsorption = Float.MIN_VALUE;
     private int lastRecordedFoodLevel = Integer.MIN_VALUE;
     private int lastRecordedAirLevel = Integer.MIN_VALUE;
@@ -492,6 +493,23 @@ public class ServerPlayer extends Player implements ca.spottedleaf.moonrise.patc
         // CraftBukkit end
     }
 
+    // MoraPaper - optimized constructor for NPCs
+    public ServerPlayer(MinecraftServer server, ServerLevel level, GameProfile gameProfile) {
+        super(level, gameProfile);
+        this.textFilter = server.createTextFilterForPlayer(this);
+        this.gameMode = server.createGameModeForPlayer(this);
+        this.server = server;
+        this.object = null;
+        // CraftBukkit start
+        this.displayName = this.getScoreboardName();
+        this.adventure$displayName = net.kyori.adventure.text.Component.text(this.getScoreboardName()); // Paper
+        this.bukkitPickUpLoot = true;
+        this.maxHealthCache = this.getMaxHealth();
+        this.recipeBook = null;
+        this.stats = null;
+        this.advancements = null;
+    }
+
     @Override
     public BlockPos adjustSpawnLocation(ServerLevel level, BlockPos pos) {
         CompletableFuture<Vec3> completableFuture = PlayerSpawnFinder.findSpawn(level, pos);
@@ -2331,7 +2349,12 @@ public class ServerPlayer extends Player implements ca.spottedleaf.moonrise.patc
         this.sendSystemMessage(mesage, false);
     }
 
+    // MoraPaper - Completely disable chat
+    public boolean chatDisabled = false;
+
     public void sendSystemMessage(Component message, boolean overlay) {
+        if (chatDisabled) return;
+
         if (this.acceptsSystemMessages(overlay)) {
             this.connection
                 .send(
@@ -2359,6 +2382,9 @@ public class ServerPlayer extends Player implements ca.spottedleaf.moonrise.patc
         this.sendChatMessage(message, filtered, boundType, null);
     }
     public void sendChatMessage(OutgoingChatMessage message, boolean filtered, ChatType.Bound boundType, @Nullable Component unsigned) {
+        // MoraPaper
+        if (chatDisabled) return;
+
         // Paper end
         if (this.acceptsChatMessages()) {
             message.sendToPlayer(this, filtered, boundType, unsigned); // Paper
diff --git a/net/minecraft/server/network/ServerGamePacketListenerImpl.java b/net/minecraft/server/network/ServerGamePacketListenerImpl.java
index 7f6cc8f111f6f5df311397a7e31e25fa2c93c685..9b5e5445b3ce3506027ef08db28d4bef6227aba4 100644
--- a/net/minecraft/server/network/ServerGamePacketListenerImpl.java
+++ b/net/minecraft/server/network/ServerGamePacketListenerImpl.java
@@ -2464,8 +2464,9 @@ public class ServerGamePacketListenerImpl
                 LastSeenMessages lastSeenMessages = this.lastSeenMessages.applyUpdate(update);
                 var10000 = Optional.of(lastSeenMessages);
             } catch (LastSeenMessagesValidator.ValidationException var5) {
-                LOGGER.error("Failed to validate message acknowledgements from {}: {}", this.player.getPlainTextName(), var5.getMessage());
-                this.disconnectAsync(CHAT_VALIDATION_FAILED, org.bukkit.event.player.PlayerKickEvent.Cause.CHAT_VALIDATION_FAILED); // Paper - kick event causes & add proper async disconnect
+                // MoraPaper - Ignore kicks for invalid messages
+                // LOGGER.error("Failed to validate message acknowledgements from {}: {}", this.player.getPlainTextName(), var5.getMessage());
+                // this.disconnectAsync(CHAT_VALIDATION_FAILED, org.bukkit.event.player.PlayerKickEvent.Cause.CHAT_VALIDATION_FAILED); // Paper - kick event causes & add proper async disconnect
                 return Optional.empty();
             }
 
@@ -2567,8 +2568,9 @@ public class ServerGamePacketListenerImpl
             try {
                 this.lastSeenMessages.applyOffset(packet.offset());
             } catch (LastSeenMessagesValidator.ValidationException var5) {
-                LOGGER.error("Failed to validate message acknowledgement offset from {}: {}", this.player.getPlainTextName(), var5.getMessage());
-                this.disconnectAsync(ServerGamePacketListenerImpl.CHAT_VALIDATION_FAILED, org.bukkit.event.player.PlayerKickEvent.Cause.CHAT_VALIDATION_FAILED); // Paper - kick event causes & add proper async disconnect
+                // MoraPaper - Ignore kicks for invalid messages
+                // LOGGER.error("Failed to validate message acknowledgement offset from {}: {}", this.player.getPlainTextName(), var5.getMessage());
+                // this.disconnectAsync(ServerGamePacketListenerImpl.CHAT_VALIDATION_FAILED, org.bukkit.event.player.PlayerKickEvent.Cause.CHAT_VALIDATION_FAILED); // Paper - kick event causes & add proper async disconnect
             }
         }
     }
@@ -2683,6 +2685,10 @@ public class ServerGamePacketListenerImpl
             return;
         }
         // CraftBukkit end
+
+        // MoraPaper
+        if (player.chatDisabled) return;
+
         // Paper start - Ensure that client receives chat packets in the same order that we add into the message signature cache
         synchronized (this.messageSignatureCache) {
         this.send(
diff --git a/net/minecraft/world/entity/LivingEntity.java b/net/minecraft/world/entity/LivingEntity.java
index 90483699b3f2efe9945c849adfdd38422e3a3458..b21d804006424dd256390b46534f2f22c1e9a36e 100644
--- a/net/minecraft/world/entity/LivingEntity.java
+++ b/net/minecraft/world/entity/LivingEntity.java
@@ -4693,4 +4693,9 @@ public abstract class LivingEntity extends Entity implements Attackable, Waypoin
 
     public record Fallsounds(SoundEvent small, SoundEvent big) {
     }
+
+    // MoraPaper start - LivingEntity equipment accessor
+    public EntityEquipment getEquipment() {
+        return this.equipment;
+    }
 }
diff --git a/net/minecraft/world/entity/Mob.java b/net/minecraft/world/entity/Mob.java
index 4994dcb209dd82d60559cfc453198b75cbfa1511..1d668404fe726916b0c091ab3999473a77d7818b 100644
--- a/net/minecraft/world/entity/Mob.java
+++ b/net/minecraft/world/entity/Mob.java
@@ -521,8 +521,13 @@ public abstract class Mob extends LivingEntity implements EquipmentUser, Leashab
         this.resetAngularLeashMomentum();
     }
 
+    // MoraPaper
+    public static int distanceSqrToDisableAi = 400;
+
     @Override
     public void aiStep() {
+        if(nearestPlayerData.distanceSqr > distanceSqrToDisableAi) return;
+
         super.aiStep();
         ProfilerFiller profilerFiller = Profiler.get();
         profilerFiller.push("looting");
@@ -719,21 +724,39 @@ public abstract class Mob extends LivingEntity implements EquipmentUser, Leashab
     }
     // Paper end - allow changing despawnInPeaceful
 
+    // MoraPaper
+    public net.minemora.MoraPaper.PlayerDistanceData nearestPlayerData = new net.minemora.MoraPaper.PlayerDistanceData();
+
+    public @Nullable Player getNearestPlayerNew() {
+        if(nearestPlayerData.uuid != null) {
+            org.bukkit.entity.Player bukkitPlayer = org.bukkit.Bukkit.getPlayer(nearestPlayerData.uuid);
+            if(bukkitPlayer != null) {
+                return ((org.bukkit.craftbukkit.entity.CraftPlayer)bukkitPlayer).getHandle();
+            }
+            else {
+                nearestPlayerData.uuid = null;
+            }
+        }
+        return null;
+    }
+
     @Override
     public void checkDespawn() {
         if (this.level().getDifficulty() == Difficulty.PEACEFUL && this.shouldDespawnInPeaceful()) { // Paper - allow changing despawnInPeaceful
             this.discard(EntityRemoveEvent.Cause.DESPAWN); // CraftBukkit - add Bukkit remove cause
         } else if (!this.isPersistenceRequired() && !this.requiresCustomPersistence()) {
-            Entity nearestPlayer = this.level().findNearbyPlayer(this, -1.0, EntitySelector.PLAYER_AFFECTS_SPAWNING); // Paper - Affects Spawning API
+            Player nearestPlayer = this.getNearestPlayerNew(); // MoraPaper - Optimize get nearest player
             if (nearestPlayer != null) {
                 // Paper start - Configurable despawn distances
+                // MoraPaper Start - Use our cached nearest player distance
                 final io.papermc.paper.configuration.WorldConfiguration.Entities.Spawning.DespawnRangePair despawnRangePair = this.level().paperConfig().entities.spawning.despawnRanges.get(this.getType().getCategory());
                 final io.papermc.paper.configuration.type.DespawnRange.Shape shape = this.level().paperConfig().entities.spawning.despawnRangeShape;
-                final double dy = Math.abs(nearestPlayer.getY() - this.getY());
-                final double dySqr = Mth.square(dy);
-                final double dxSqr = Mth.square(nearestPlayer.getX() - this.getX());
-                final double dzSqr = Mth.square(nearestPlayer.getZ() - this.getZ());
-                final double distanceSquared = dxSqr + dzSqr + dySqr;
+                final double dy = nearestPlayerData.dy;
+                final double dySqr = nearestPlayerData.dySqr;
+                final double dxSqr = nearestPlayerData.dxSqr;
+                final double dzSqr = nearestPlayerData.dzSqr;
+                final double distanceSquared = nearestPlayerData.distanceSqr;
+                // MoraPaper End
                 // Despawn if hard/soft limit is exceeded
                 if (despawnRangePair.hard().shouldDespawn(shape, dxSqr, dySqr, dzSqr, dy) && this.removeWhenFarAway(distanceSquared)) {
                     this.discard(EntityRemoveEvent.Cause.DESPAWN); // CraftBukkit - add Bukkit remove cause
diff --git a/net/minecraft/world/entity/ai/goal/GoalSelector.java b/net/minecraft/world/entity/ai/goal/GoalSelector.java
index 859b859d29b637200cf7c9a0bd52d9f712413e3d..b77c5b992fb09518f3da352af0a0a424b1e40b10 100644
--- a/net/minecraft/world/entity/ai/goal/GoalSelector.java
+++ b/net/minecraft/world/entity/ai/goal/GoalSelector.java
@@ -38,7 +38,7 @@ public class GoalSelector {
     // Paper start - EAR 2
     public boolean inactiveTick() {
         this.curRate++;
-        return this.curRate % 3 == 0; // TODO newGoalRate was already unused in 1.20.4, check if this is correct
+        return this.curRate % 40 == 0; // MoraPaper - Increase curRate from 3 to 40
     }
 
     public boolean hasTasks() {
diff --git a/net/minecraft/world/entity/boss/enderdragon/EnderDragon.java b/net/minecraft/world/entity/boss/enderdragon/EnderDragon.java
index 0c900fd908792971180922a386768ec561e10e6e..8c852f3a0ee6283dc612a43c2108fc8a933d1057 100644
--- a/net/minecraft/world/entity/boss/enderdragon/EnderDragon.java
+++ b/net/minecraft/world/entity/boss/enderdragon/EnderDragon.java
@@ -674,6 +674,28 @@ public class EnderDragon extends Mob implements Enemy {
         }
     }
 
+    protected float getFirstSearchRadius() {
+        return 60.0F;
+    }
+
+    protected float getSecondSearchRadius() {
+        return 40.0F;
+    }
+
+    protected float getThirdSearchRadius() {
+        return 20.0F;
+    }
+
+    protected int calculateMaxY(BlockPos position, int i1) {
+        return Math.max(73, this.level().getHeightmapPos(Heightmap.Types.MOTION_BLOCKING_NO_LEAVES, position).getY() + i1);
+    }
+
+    // MoraPaper - Add the offset to the vector based on origin
+    public void addOrigin(Vec3 vec3) {
+        final net.minecraft.core.BlockPos fightOrigin = getFightOrigin();
+        vec3.add(fightOrigin.getX(), 0, fightOrigin.getZ());
+    }
+
     public int findClosestNode() {
         if (this.nodes[0] == null) {
             for (int i = 0; i < 24; i++) {
@@ -681,21 +703,25 @@ public class EnderDragon extends Mob implements Enemy {
                 int floor;
                 int floor1;
                 if (i < 12) {
-                    floor = Mth.floor(60.0F * Mth.cos(2.0F * ((float) -Math.PI + (float) (Math.PI / 12) * i)));
-                    floor1 = Mth.floor(60.0F * Mth.sin(2.0F * ((float) -Math.PI + (float) (Math.PI / 12) * i)));
+                    floor = Mth.floor(getFirstSearchRadius() * Mth.cos(2.0F * ((float) -Math.PI + (float) (Math.PI / 12) * i)));
+                    floor1 = Mth.floor(getFirstSearchRadius() * Mth.sin(2.0F * ((float) -Math.PI + (float) (Math.PI / 12) * i)));
                 } else if (i < 20) {
                     int i2 = i - 12;
-                    floor = Mth.floor(40.0F * Mth.cos(2.0F * ((float) -Math.PI + (float) (Math.PI / 8) * i2)));
-                    floor1 = Mth.floor(40.0F * Mth.sin(2.0F * ((float) -Math.PI + (float) (Math.PI / 8) * i2)));
+                    floor = Mth.floor(getSecondSearchRadius() * Mth.cos(2.0F * ((float) -Math.PI + (float) (Math.PI / 8) * i2)));
+                    floor1 = Mth.floor(getSecondSearchRadius() * Mth.sin(2.0F * ((float) -Math.PI + (float) (Math.PI / 8) * i2)));
                     i1 += 10;
                 } else {
                     int var7 = i - 20;
-                    floor = Mth.floor(20.0F * Mth.cos(2.0F * ((float) -Math.PI + (float) (Math.PI / 4) * var7)));
-                    floor1 = Mth.floor(20.0F * Mth.sin(2.0F * ((float) -Math.PI + (float) (Math.PI / 4) * var7)));
+                    floor = Mth.floor(getThirdSearchRadius() * Mth.cos(2.0F * ((float) -Math.PI + (float) (Math.PI / 4) * var7)));
+                    floor1 = Mth.floor(getThirdSearchRadius() * Mth.sin(2.0F * ((float) -Math.PI + (float) (Math.PI / 4) * var7)));
                 }
 
-                int max = Math.max(73, this.level().getHeightmapPos(Heightmap.Types.MOTION_BLOCKING_NO_LEAVES, new BlockPos(floor, 0, floor1)).getY() + i1);
-                this.nodes[i] = new Node(floor, max, floor1);
+                // MoraPaper - Offset nodes to origin
+                floor += getFightOrigin().getX();
+                floor1 += getFightOrigin().getZ();
+
+                int maxY = calculateMaxY(new BlockPos(floor, 0, floor1), i1);
+                this.nodes[i] = new Node(floor, maxY, floor1);
             }
 
             this.nodeAdjacency[0] = 6146;
@@ -900,7 +926,7 @@ public class EnderDragon extends Mob implements Enemy {
         EnderDragonPhase<? extends DragonPhaseInstance> phase = currentPhase.getPhase();
         Vec3 viewVector;
         if (phase == EnderDragonPhase.LANDING || phase == EnderDragonPhase.TAKEOFF) {
-            BlockPos heightmapPos = this.level().getHeightmapPos(Heightmap.Types.MOTION_BLOCKING_NO_LEAVES, this.getPodium()); // Paper - Allow changing the EnderDragon podium
+            BlockPos heightmapPos = calculateLandPosition((net.minecraft.server.level.ServerLevel) this.level(), net.minecraft.world.level.levelgen.Heightmap.Types.MOTION_BLOCKING_NO_LEAVES); // MoraPaper - Allow changing the EnderDragon podium
             float max = Math.max((float)Math.sqrt(heightmapPos.distToCenterSqr(this.position())) / 4.0F, 1.0F);
             float f = 6.0F / max;
             float xRot = this.getXRot();
@@ -1002,5 +1028,11 @@ public class EnderDragon extends Mob implements Enemy {
 
         return flag ? i : 0;
     }
+
+    // MoraSpigot - Allow overriding of land position to ignore the heightmap position
+    public net.minecraft.core.BlockPos calculateLandPosition(final net.minecraft.server.level.ServerLevel level, final net.minecraft.world.level.levelgen.Heightmap.Types heightmapType) {
+        return level.getHeightmapPos(heightmapType, getPodium()); // Paper - Allow changing the EnderDragon podium
+    }
+
     // CraftBukkit end
 }
diff --git a/net/minecraft/world/entity/boss/enderdragon/phases/DragonDeathPhase.java b/net/minecraft/world/entity/boss/enderdragon/phases/DragonDeathPhase.java
index 85e20366beb7c44c938aa1e9bc2a806ddfa173a0..4bb72708cbc92274451c52e2d44157ebd5ed313b 100644
--- a/net/minecraft/world/entity/boss/enderdragon/phases/DragonDeathPhase.java
+++ b/net/minecraft/world/entity/boss/enderdragon/phases/DragonDeathPhase.java
@@ -6,7 +6,6 @@ import net.minecraft.core.particles.ParticleTypes;
 import net.minecraft.server.level.ServerLevel;
 import net.minecraft.world.entity.boss.enderdragon.EnderDragon;
 import net.minecraft.world.level.levelgen.Heightmap;
-import net.minecraft.world.level.levelgen.feature.EndPodiumFeature;
 import net.minecraft.world.phys.Vec3;
 
 public class DragonDeathPhase extends AbstractDragonPhaseInstance {
@@ -34,7 +33,7 @@ public class DragonDeathPhase extends AbstractDragonPhaseInstance {
     public void doServerTick(ServerLevel level) {
         this.time++;
         if (this.targetLocation == null) {
-            BlockPos heightmapPos = level.getHeightmapPos(Heightmap.Types.MOTION_BLOCKING, this.dragon.getPodium()); // Paper - Allow changing the EnderDragon podium
+            BlockPos heightmapPos = this.dragon.calculateLandPosition(level, Heightmap.Types.MOTION_BLOCKING); // MoraPaper - Allow changing the EnderDragon podium
             this.targetLocation = Vec3.atBottomCenterOf(heightmapPos);
         }
 
diff --git a/net/minecraft/world/entity/boss/enderdragon/phases/DragonHoldingPatternPhase.java b/net/minecraft/world/entity/boss/enderdragon/phases/DragonHoldingPatternPhase.java
index 1afb9ff6a5546f912238b1b03aacf45033f7c481..51b14c7f52f1a46f7452ced8e1966ae0ec3df5a1 100644
--- a/net/minecraft/world/entity/boss/enderdragon/phases/DragonHoldingPatternPhase.java
+++ b/net/minecraft/world/entity/boss/enderdragon/phases/DragonHoldingPatternPhase.java
@@ -53,7 +53,7 @@ public class DragonHoldingPatternPhase extends AbstractDragonPhaseInstance {
 
     private void findNewTarget(ServerLevel level) {
         if (this.currentPath != null && this.currentPath.isDone()) {
-            BlockPos heightmapPos = level.getHeightmapPos(Heightmap.Types.MOTION_BLOCKING_NO_LEAVES, this.dragon.getPodium()); // Paper - Allow changing the EnderDragon podium
+            BlockPos heightmapPos = this.dragon.calculateLandPosition(level, Heightmap.Types.MOTION_BLOCKING_NO_LEAVES); // MoraPaper - Allow changing the EnderDragon podium
             int i = this.dragon.getDragonFight() == null ? 0 : this.dragon.getDragonFight().getCrystalsAlive();
             if (this.dragon.getRandom().nextInt(i + 3) == 0) {
                 this.dragon.getPhaseManager().setPhase(EnderDragonPhase.LANDING_APPROACH);
diff --git a/net/minecraft/world/entity/boss/enderdragon/phases/DragonLandingApproachPhase.java b/net/minecraft/world/entity/boss/enderdragon/phases/DragonLandingApproachPhase.java
index aafa99c3553a1ba35e3d7e1a7c7842ed0468e509..a2e0314e52c7188e086b4dbf80285a750ebdbfee 100644
--- a/net/minecraft/world/entity/boss/enderdragon/phases/DragonLandingApproachPhase.java
+++ b/net/minecraft/world/entity/boss/enderdragon/phases/DragonLandingApproachPhase.java
@@ -7,8 +7,6 @@ import net.minecraft.server.level.ServerLevel;
 import net.minecraft.world.entity.ai.targeting.TargetingConditions;
 import net.minecraft.world.entity.boss.enderdragon.EnderDragon;
 import net.minecraft.world.entity.player.Player;
-import net.minecraft.world.level.levelgen.Heightmap;
-import net.minecraft.world.level.levelgen.feature.EndPodiumFeature;
 import net.minecraft.world.level.pathfinder.Node;
 import net.minecraft.world.level.pathfinder.Path;
 import net.minecraft.world.phys.Vec3;
@@ -49,17 +47,28 @@ public class DragonLandingApproachPhase extends AbstractDragonPhaseInstance {
         return this.targetLocation;
     }
 
+
+
     private void findNewTarget(ServerLevel level) {
         if (this.currentPath == null || this.currentPath.isDone()) {
             int i = this.dragon.findClosestNode();
-            BlockPos heightmapPos = level.getHeightmapPos(Heightmap.Types.MOTION_BLOCKING_NO_LEAVES, this.dragon.getPodium()); // Paper - Allow changing the EnderDragon podium
+            BlockPos heightmapPos = this.dragon.calculateLandPosition(level, net.minecraft.world.level.levelgen.Heightmap.Types.MOTION_BLOCKING_NO_LEAVES); // MoraPaper - Allow changing the EnderDragon podium
             Player nearestPlayer = level.getNearestPlayer(NEAR_EGG_TARGETING, this.dragon, heightmapPos.getX(), heightmapPos.getY(), heightmapPos.getZ());
             int i1;
             if (nearestPlayer != null) {
                 Vec3 vec3 = new Vec3(nearestPlayer.getX(), 0.0, nearestPlayer.getZ()).normalize();
-                i1 = this.dragon.findClosestNode(-vec3.x * 40.0, 105.0, -vec3.z * 40.0);
+
+                // MoraPaper - Add origin
+                Vec3 extra = new Vec3(-vec3.x * 40.0, 105.0, -vec3.z * 40.0);
+                this.dragon.addOrigin(extra);
+
+                i1 = this.dragon.findClosestNode(extra.x, extra.y, extra.z);
             } else {
-                i1 = this.dragon.findClosestNode(40.0, heightmapPos.getY(), 0.0);
+                // MoraPaper - Add origin
+                Vec3 extra = new Vec3(40.0, heightmapPos.getY(), 0.0);
+                this.dragon.addOrigin(extra);
+
+                i1 = this.dragon.findClosestNode(extra.x, extra.y, extra.z);
             }
 
             Node node = new Node(heightmapPos.getX(), heightmapPos.getY(), heightmapPos.getZ());
diff --git a/net/minecraft/world/entity/boss/enderdragon/phases/DragonLandingPhase.java b/net/minecraft/world/entity/boss/enderdragon/phases/DragonLandingPhase.java
index d3c9304845f3bd517a2188cb5d3ae7561b530d17..629fd3ca8767a595651f06142f19830c0f8c9cfb 100644
--- a/net/minecraft/world/entity/boss/enderdragon/phases/DragonLandingPhase.java
+++ b/net/minecraft/world/entity/boss/enderdragon/phases/DragonLandingPhase.java
@@ -51,7 +51,7 @@ public class DragonLandingPhase extends AbstractDragonPhaseInstance {
     public void doServerTick(ServerLevel level) {
         if (this.targetLocation == null) {
             this.targetLocation = Vec3.atBottomCenterOf(
-                level.getHeightmapPos(Heightmap.Types.MOTION_BLOCKING_NO_LEAVES, this.dragon.getPodium()) // Paper - Allow changing the EnderDragon podium
+                this.dragon.calculateLandPosition(level, net.minecraft.world.level.levelgen.Heightmap.Types.MOTION_BLOCKING_NO_LEAVES)  // MoraPaper - Allow changing the EnderDragon podium
             );
         }
 
diff --git a/net/minecraft/world/entity/boss/enderdragon/phases/DragonTakeoffPhase.java b/net/minecraft/world/entity/boss/enderdragon/phases/DragonTakeoffPhase.java
index d0142ec67dad3b41ae0a693df614956e8ddabd9c..701873f0e84a07eeb81a9a7e70ec4967f27837d3 100644
--- a/net/minecraft/world/entity/boss/enderdragon/phases/DragonTakeoffPhase.java
+++ b/net/minecraft/world/entity/boss/enderdragon/phases/DragonTakeoffPhase.java
@@ -24,7 +24,7 @@ public class DragonTakeoffPhase extends AbstractDragonPhaseInstance {
     @Override
     public void doServerTick(ServerLevel level) {
         if (!this.firstTick && this.currentPath != null) {
-            BlockPos heightmapPos = level.getHeightmapPos(Heightmap.Types.MOTION_BLOCKING_NO_LEAVES, this.dragon.getPodium()); // Paper - Allow changing the EnderDragon podium
+            BlockPos heightmapPos = this.dragon.calculateLandPosition(level, Heightmap.Types.MOTION_BLOCKING_NO_LEAVES); // MoraPaper - Allow changing the EnderDragon podium
             if (!heightmapPos.closerToCenterThan(this.dragon.position(), 10.0)) {
                 this.dragon.getPhaseManager().setPhase(EnderDragonPhase.HOLDING_PATTERN);
             }
@@ -44,7 +44,12 @@ public class DragonTakeoffPhase extends AbstractDragonPhaseInstance {
     private void findNewTarget() {
         int i = this.dragon.findClosestNode();
         Vec3 headLookVector = this.dragon.getHeadLookVector(1.0F);
-        int i1 = this.dragon.findClosestNode(-headLookVector.x * 40.0, 105.0, -headLookVector.z * 40.0);
+
+        // MoraPaper - Add origin
+        Vec3 extra = new Vec3(-headLookVector.x * 40.0, 105.0, -headLookVector.z * 40.0);
+        this.dragon.addOrigin(extra);
+
+        int i1 = this.dragon.findClosestNode();
         if (this.dragon.getDragonFight() != null && this.dragon.getDragonFight().getCrystalsAlive() > 0) {
             i1 %= 12;
             if (i1 < 0) {
diff --git a/net/minecraft/world/entity/monster/Monster.java b/net/minecraft/world/entity/monster/Monster.java
index 8cbef9c47f7022e9b90b0bf2718382355d74a686..7c3405b2462c5382d24041f9fb9885e8d56ed1eb 100644
--- a/net/minecraft/world/entity/monster/Monster.java
+++ b/net/minecraft/world/entity/monster/Monster.java
@@ -41,6 +41,7 @@ public abstract class Monster extends PathfinderMob implements Enemy {
 
     @Override
     public void aiStep() {
+        if (nearestPlayerData.distanceSqr > distanceSqrToDisableAi) return;
         this.updateSwingTime();
         this.updateNoActionTime();
         super.aiStep();
diff --git a/net/minecraft/world/entity/npc/Villager.java b/net/minecraft/world/entity/npc/Villager.java
index ae2f000083556cdc1cbb637e6210c4c4e22c974f..aa39f787a1dade97155d03566848e1104a254f43 100644
--- a/net/minecraft/world/entity/npc/Villager.java
+++ b/net/minecraft/world/entity/npc/Villager.java
@@ -292,7 +292,12 @@ public class Villager extends AbstractVillager implements ReputationEventHandler
         // Paper start - EAR 2
         this.customServerAiStep(level, false);
     }
-    protected void customServerAiStep(ServerLevel level, final boolean inactive) {
+    protected void customServerAiStep(ServerLevel level, boolean inactive) {
+        // MoraSpigot - Mark inactive if the player is too far away
+        if(nearestPlayerData.distanceSqr > distanceSqrToDisableAi) {
+            inactive = true;
+        }
+
         // Paper end - EAR 2
         ProfilerFiller profilerFiller = Profiler.get();
         profilerFiller.push("villagerBrain");
diff --git a/net/minecraft/world/entity/projectile/Snowball.java b/net/minecraft/world/entity/projectile/Snowball.java
index 52537b157604d305e2a3db958dc5ada852a148b5..207a7f6b3c1d7f2451f76edc90a5f3b1a4ddeed5 100644
--- a/net/minecraft/world/entity/projectile/Snowball.java
+++ b/net/minecraft/world/entity/projectile/Snowball.java
@@ -64,4 +64,13 @@ public class Snowball extends ThrowableItemProjectile {
             this.discard(org.bukkit.event.entity.EntityRemoveEvent.Cause.HIT); // CraftBukkit - add Bukkit remove cause
         }
     }
+
+    // MoraPaper start - Prevent Snowball Lag Machine
+    @Override
+    public void tick() {
+        super.tick();
+        if (this.tickCount >= 8) discard();
+    }
+    // MoraPaper end - Prevent Snowball Lag Machine
+
 }
diff --git a/net/minecraft/world/entity/projectile/ThrownEgg.java b/net/minecraft/world/entity/projectile/ThrownEgg.java
index 57cb5b73dae86379be857ffa0f55ef7a056a0b4a..49b7f0a06521ec7c7a411976afa9aebcb6d1de3b 100644
--- a/net/minecraft/world/entity/projectile/ThrownEgg.java
+++ b/net/minecraft/world/entity/projectile/ThrownEgg.java
@@ -128,4 +128,13 @@ public class ThrownEgg extends ThrowableItemProjectile {
     public Item getDefaultItem() {
         return Items.EGG;
     }
+
+    // MoraPaper start - Prevent Egg Lag Machine
+    @Override
+    public void tick() {
+        super.tick();
+        if (this.tickCount >= 8) discard();
+    }
+    // MoraPaper end - Prevent Egg Lag Machine
+
 }
diff --git a/net/minecraft/world/entity/vehicle/NewMinecartBehavior.java b/net/minecraft/world/entity/vehicle/NewMinecartBehavior.java
index 090b19752fbfc856d9fbf118510afc6cda2b9989..71486d7a2c7b7e775db283f937c9ffce1268d4db 100644
--- a/net/minecraft/world/entity/vehicle/NewMinecartBehavior.java
+++ b/net/minecraft/world/entity/vehicle/NewMinecartBehavior.java
@@ -367,6 +367,13 @@ public class NewMinecartBehavior extends MinecartBehavior {
         if (this.minecart.getFirstPassenger() instanceof ServerPlayer serverPlayer) {
             Vec3 lastClientMoveIntent = serverPlayer.getLastClientMoveIntent();
             if (lastClientMoveIntent.lengthSqr() > 0.0) {
+                // MoraPaper Start - Steer minecart event
+                final org.bukkit.event.player.PlayerSteerMinecartEvent event = new org.bukkit.event.player.PlayerSteerMinecartEvent(serverPlayer.getBukkitEntity(), (org.bukkit.entity.Minecart) this.minecart.getBukkitEntity());
+                event.callEvent();
+
+                if (event.isCancelled()) return speed;
+                // MoraPaper End
+
                 Vec3 vec3 = lastClientMoveIntent.normalize();
                 double d = speed.horizontalDistanceSqr();
                 if (vec3.lengthSqr() > 0.0 && d < 0.01) {
diff --git a/net/minecraft/world/entity/vehicle/OldMinecartBehavior.java b/net/minecraft/world/entity/vehicle/OldMinecartBehavior.java
index bf91da8b6a5ff33af1e9a8d2271466df1a129f20..3fc875114e349951cb9bec7d0e8f7ff04c0f4d8e 100644
--- a/net/minecraft/world/entity/vehicle/OldMinecartBehavior.java
+++ b/net/minecraft/world/entity/vehicle/OldMinecartBehavior.java
@@ -162,6 +162,17 @@ public class OldMinecartBehavior extends MinecartBehavior {
         Vec3 lastClientMoveIntent;
         if (this.minecart.getFirstPassenger() instanceof ServerPlayer serverPlayer) {
             lastClientMoveIntent = serverPlayer.getLastClientMoveIntent();
+
+            // MoraPaper Start - Steer minecart event
+            if (lastClientMoveIntent.x != 0 || lastClientMoveIntent.y != 0 || lastClientMoveIntent.z != 0) {
+                final org.bukkit.event.player.PlayerSteerMinecartEvent event = new org.bukkit.event.player.PlayerSteerMinecartEvent(serverPlayer.getBukkitEntity(), (org.bukkit.entity.Minecart) this.minecart.getBukkitEntity());
+                event.callEvent();
+
+                if (event.isCancelled()) {
+                    lastClientMoveIntent = net.minecraft.world.phys.Vec3.ZERO;
+                }
+            }
+            // MoraPaper End
         } else {
             lastClientMoveIntent = Vec3.ZERO;
         }
diff --git a/net/minecraft/world/level/entity/EntityTickList.java b/net/minecraft/world/level/entity/EntityTickList.java
index 423779a2b690f387a4f0bd07b97b50e0baefda76..33cdab242a3d58c6cbe4ede3497f1864f05bcf7e 100644
--- a/net/minecraft/world/level/entity/EntityTickList.java
+++ b/net/minecraft/world/level/entity/EntityTickList.java
@@ -11,6 +11,9 @@ import net.minecraft.world.entity.Entity;
 public class EntityTickList {
     private final ca.spottedleaf.moonrise.common.list.IteratorSafeOrderedReferenceSet<net.minecraft.world.entity.Entity> entities = new ca.spottedleaf.moonrise.common.list.IteratorSafeOrderedReferenceSet<>(); // Paper - rewrite chunk system
 
+    // MoraPaper - async list to use entities from other threads
+    private final java.util.concurrent.ConcurrentLinkedQueue<net.minecraft.world.entity.Entity> asyncEntities = new java.util.concurrent.ConcurrentLinkedQueue<>();
+
     private void ensureActiveIsNotIterated() {
         // Paper - rewrite chunk system
     }
@@ -18,11 +21,13 @@ public class EntityTickList {
     public void add(Entity entity) {
         this.ensureActiveIsNotIterated();
         this.entities.add(entity); // Paper - rewrite chunk system
+        this.asyncEntities.add(entity); // MoraPaper - async list to use entities from other threads
     }
 
     public void remove(Entity entity) {
         this.ensureActiveIsNotIterated();
         this.entities.remove(entity); // Paper - rewrite chunk system
+        this.asyncEntities.remove(entity); // MoraPaper - async list to use entities from other threads
     }
 
     public boolean contains(Entity entity) {
@@ -43,4 +48,10 @@ public class EntityTickList {
         }
         // Paper end - rewrite chunk system
     }
+
+    // MoraPaper - async list to use entities from other threads
+    public java.util.concurrent.ConcurrentLinkedQueue<Entity> getEntitiesAsync() {
+        return asyncEntities;
+    }
+
 }
