From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Tehin <62246450+Tehin@users.noreply.github.com>
Date: Wed, 2 Apr 2025 22:23:25 -0500
Subject: [PATCH] Minemora Patch


diff --git a/ca/spottedleaf/moonrise/patches/chunk_system/scheduling/ChunkHolderManager.java b/ca/spottedleaf/moonrise/patches/chunk_system/scheduling/ChunkHolderManager.java
index 05d374f679919d082064e8f2bb363a419bc9648e..7d9f48fbe2456ebb616468a8e79bfaaf4fe3e1ab 100644
--- a/ca/spottedleaf/moonrise/patches/chunk_system/scheduling/ChunkHolderManager.java
+++ b/ca/spottedleaf/moonrise/patches/chunk_system/scheduling/ChunkHolderManager.java
@@ -70,6 +70,9 @@ public final class ChunkHolderManager {
     public static final int ENTITY_TICKING_TICKET_LEVEL = ChunkLevel.ENTITY_TICKING_LEVEL;
     public static final int MAX_TICKET_LEVEL = ChunkLevel.MAX_LEVEL; // inclusive
 
+    // MoraPaper - Limit chunk unloads per tick
+    private static final int MAX_CHUNK_UNLOADS_PER_TICK = 180;
+
     public static final TicketType UNLOAD_COOLDOWN = ChunkSystemTicketType.create("chunk_system:unload_cooldown", null, 5L * 20L);
 
     private static final long NO_TIMEOUT_MARKER = Long.MIN_VALUE;
@@ -1181,10 +1184,8 @@ public final class ChunkHolderManager {
         // We do need to process updates here so that any addTicket that is synchronised before this call does not go missed.
         this.processTicketUpdates();
 
-        final int toUnloadCount = Math.max(
-            PlatformHooks.get().configMinChunkUnloadCount(this.world),
-            (int)Math.round(unloadCountTentative * Math.clamp(PlatformHooks.get().configMinChunkUnloadFraction(this.world), 0.0, 1.0))
-        );
+        // MoraPaper - Limit chunk unload to prevent RAM overuse
+        final int toUnloadCount = Math.max(50,Math.min(MAX_CHUNK_UNLOADS_PER_TICK, (int) (unloadCountTentative * 0.05)));
         int processedCount = 0;
 
         for (final ChunkUnloadQueue.SectionToUnload sectionRef : unloadSectionsForRegion) {
diff --git a/net/minecraft/network/chat/OutgoingChatMessage.java b/net/minecraft/network/chat/OutgoingChatMessage.java
index 259cbcbdd3cd47907629772dc7565212bb434729..e10c71cab9ebbc48b5303bb832672c42aa2ad9cc 100644
--- a/net/minecraft/network/chat/OutgoingChatMessage.java
+++ b/net/minecraft/network/chat/OutgoingChatMessage.java
@@ -25,6 +25,9 @@ public interface OutgoingChatMessage {
         }
         @Override
         public void sendToPlayer(final ServerPlayer player, final boolean filtered, final ChatType.Bound chatType, final @org.jspecify.annotations.Nullable Component unsigned) {
+            // MoraPaper
+            if (player.chatDisabled) return;
+
             player.connection.sendDisguisedChatMessage(unsigned != null ? unsigned : this.content, chatType);
             // Paper end
         }
@@ -43,6 +46,9 @@ public interface OutgoingChatMessage {
         }
         @Override
         public void sendToPlayer(final ServerPlayer player, final boolean filtered, final ChatType.Bound chatType, final @org.jspecify.annotations.Nullable Component unsigned) {
+            // MoraPaper
+            if (player.chatDisabled) return;
+
             // Paper end
             PlayerChatMessage filteredMessage = this.message.filter(filtered);
             filteredMessage = unsigned != null ? filteredMessage.withUnsignedContent(unsigned) : filteredMessage; // Paper
diff --git a/net/minecraft/network/protocol/game/ClientboundPlayerInfoUpdatePacket.java b/net/minecraft/network/protocol/game/ClientboundPlayerInfoUpdatePacket.java
index aaa497096f07f2ef9da62b9760f5c9aa0b0263d3..58cfca970db9b2f045e00cc012741208fbdab5b1 100644
--- a/net/minecraft/network/protocol/game/ClientboundPlayerInfoUpdatePacket.java
+++ b/net/minecraft/network/protocol/game/ClientboundPlayerInfoUpdatePacket.java
@@ -226,7 +226,7 @@ public class ClientboundPlayerInfoUpdatePacket implements Packet<ClientGamePacke
                 player.getUUID(),
                 player.getGameProfile(),
                 listed, // Paper - Add Listing API for Player
-                player.connection.latency(),
+                player.connection == null ? player.fakePing : player.connection.latency(),
                 player.gameMode(),
                 player.getTabListDisplayName(),
                 player.isModelPartShown(PlayerModelPart.HAT),
diff --git a/net/minecraft/server/MinecraftServer.java b/net/minecraft/server/MinecraftServer.java
index cc35c6271c98615a486422d0490145227626bd85..ea01a1cf0bf4c8edf7d9a500eeee98b119f82c9d 100644
--- a/net/minecraft/server/MinecraftServer.java
+++ b/net/minecraft/server/MinecraftServer.java
@@ -1791,14 +1791,14 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
         this.getPlayerList().getPlayers().forEach(playerx -> playerx.connection.suspendFlushing());
         this.server.getScheduler().mainThreadHeartbeat(); // CraftBukkit
         // Paper start - optimise Folia entity scheduler
-        ((io.papermc.paper.threadedregions.scheduler.FoliaGlobalRegionScheduler)org.bukkit.Bukkit.getGlobalRegionScheduler()).tick();
-        for (io.papermc.paper.threadedregions.EntityScheduler scheduler : this.entitySchedulerTickList.getAllSchedulers()) {
-            if (scheduler.isRetired()) {
-                continue;
-            }
-
-            scheduler.executeTick();
-        }
+        // ((io.papermc.paper.threadedregions.scheduler.FoliaGlobalRegionScheduler)org.bukkit.Bukkit.getGlobalRegionScheduler()).tick();
+        // for (io.papermc.paper.threadedregions.EntityScheduler scheduler : this.entitySchedulerTickList.getAllSchedulers()) {
+        //     if (scheduler.isRetired()) {
+        //         continue;
+        //     }
+
+        //     scheduler.executeTick();
+        // }
         // Paper end - optimise Folia entity scheduler
         // Paper end - Folia scheduler API
         io.papermc.paper.adventure.providers.ClickCallbackProviderImpl.ADVENTURE_CLICK_MANAGER.handleQueue(this.tickCount); // Paper
diff --git a/net/minecraft/server/level/ChunkMap.java b/net/minecraft/server/level/ChunkMap.java
index 42ec67c53c35189838ce4767c2be620e05108159..cc714ddf92886db013fa83f96e7f598a6862f7f3 100644
--- a/net/minecraft/server/level/ChunkMap.java
+++ b/net/minecraft/server/level/ChunkMap.java
@@ -1242,11 +1242,16 @@ public class ChunkMap extends SimpleRegionStorage implements ChunkHolder.PlayerP
             }
 
             if (lastChunkUpdate != currChunkUpdate || lastTrackedChunk != chunk) {
+                // MoraPaper - Iterate in the set itself not by creating a completely new arraylist
+                final java.util.Iterator<net.minecraft.server.network.ServerPlayerConnection> seenByIterator = this.seenBy.iterator();
+
                 // need to purge any players possible not in the chunk list
-                for (final ServerPlayerConnection conn : new java.util.ArrayList<>(this.seenBy)) {
+                while (seenByIterator.hasNext()) {
+                    final ServerPlayerConnection conn = seenByIterator.next();
                     final ServerPlayer player = conn.getPlayer();
                     if (!players.contains(player)) {
-                        this.removePlayer(player);
+                        seenByIterator.remove();
+                        this.forceRemovePlayerPairing(player);
                     }
                 }
             }
@@ -1350,6 +1355,12 @@ public class ChunkMap extends SimpleRegionStorage implements ChunkHolder.PlayerP
             }
         }
 
+        // MoraSpigot - Remove the player without removing or checking for the seenBy, to prevent CME
+        // when the player is removed while in a loop
+        public void forceRemovePlayerPairing(ServerPlayer player) {
+            this.serverEntity.removePairing(player);
+        }
+
         public void updatePlayer(final ServerPlayer player) {
             org.spigotmc.AsyncCatcher.catchOp("player tracker update"); // Spigot
             if (player != this.entity) {
diff --git a/net/minecraft/server/level/ServerChunkCache.java b/net/minecraft/server/level/ServerChunkCache.java
index 396ab3cdcf6f726ce7133395bb39594ef4b1c011..8afcdca53c2ac08da808f732f63e4cee88ac3421 100644
--- a/net/minecraft/server/level/ServerChunkCache.java
+++ b/net/minecraft/server/level/ServerChunkCache.java
@@ -128,9 +128,11 @@ public class ServerChunkCache extends ChunkSource implements ca.spottedleaf.moon
         return ret;
     }
 
-    private ChunkAccess getChunkFallback(final int chunkX, final int chunkZ, final ChunkStatus toStatus,
-                                         final boolean load) {
-        final ca.spottedleaf.moonrise.patches.chunk_system.scheduling.ChunkTaskScheduler chunkTaskScheduler = ((ca.spottedleaf.moonrise.patches.chunk_system.level.ChunkSystemServerLevel)this.level).moonrise$getChunkTaskScheduler();
+    private ChunkAccess getChunkFallback(
+        final int chunkX, final int chunkZ, final ChunkStatus toStatus,
+        final boolean load
+    ) {
+        final ca.spottedleaf.moonrise.patches.chunk_system.scheduling.ChunkTaskScheduler chunkTaskScheduler = ((ca.spottedleaf.moonrise.patches.chunk_system.level.ChunkSystemServerLevel) this.level).moonrise$getChunkTaskScheduler();
         final ca.spottedleaf.moonrise.patches.chunk_system.scheduling.ChunkHolderManager chunkHolderManager = chunkTaskScheduler.chunkHolderManager;
 
         final ca.spottedleaf.moonrise.patches.chunk_system.scheduling.NewChunkHolder currentChunk = chunkHolderManager.getChunkHolder(ca.spottedleaf.moonrise.common.util.CoordinateUtils.getChunkKey(chunkX, chunkZ));
@@ -532,7 +534,23 @@ public class ServerChunkCache extends ChunkSource implements ca.spottedleaf.moon
         profiler.pop();
     }
 
+    // MoraPaper - Only do chunk mob spawning each 5 ticks instead of every tick
+    private final static int SPAWN_UPDATE_INTERVAL = 5;
+
+    // MoraPaper - The last time we ticked mob spawning
+    private int ticksSinceLastSpawnUpdate = 0;
+
     private void tickChunks(final ProfilerFiller profiler, final long timeDiff) {
+        ticksSinceLastSpawnUpdate++;
+
+        // MoraPaper - Tick without the mob spawning between the interval
+        if (ticksSinceLastSpawnUpdate <= SPAWN_UPDATE_INTERVAL) {
+            iterateTickingChunksFaster();
+            return;
+        }
+
+        ticksSinceLastSpawnUpdate = 0;
+
         profiler.push("naturalSpawnCount");
         int chunkCount = this.distanceManager.getNaturalSpawnChunkCount();
         // Paper start - Optional per player mob spawns
@@ -615,12 +633,21 @@ public class ServerChunkCache extends ChunkSource implements ca.spottedleaf.moon
         }
 
         if (!spawningCategories.isEmpty()) {
-            if (this.level.getWorldBorder().isWithinBounds(chunkPos)) { // Paper - rewrite chunk system
+            if (this.level.getWorldBorder().isWithinBounds(chunkPos) && !isInNoNaturalSpawnRadius(chunkPos)) { // Paper - rewrite chunk system
                 NaturalSpawner.spawnForChunk(this.level, chunk, spawnCookie, spawningCategories);
             }
         }
     }
 
+    // MoraPaper Start - Limit natural spawn radius
+    public static boolean isInNoNaturalSpawnRadius(ChunkPos pos) {
+        int radius = 10; //TODO from config
+        int x = pos.x();
+        int z = pos.z();
+        return x >= -radius && x < radius && z >= -radius && z < radius;
+    }
+    // MoraPaper End
+
     private void getFullChunk(final long chunkKey, final Consumer<LevelChunk> output) {
         // Paper start - rewrite chunk system
         // note: bypass currentlyLoaded from getChunkNow
diff --git a/net/minecraft/server/level/ServerLevel.java b/net/minecraft/server/level/ServerLevel.java
index 229ff740e0418a05d824a67401ae6f105b4c3642..7dec51081772e8ae9d67fe912728f9958cadc2e7 100644
--- a/net/minecraft/server/level/ServerLevel.java
+++ b/net/minecraft/server/level/ServerLevel.java
@@ -191,7 +191,7 @@ public class ServerLevel extends Level implements WorldGenLevel, ServerEntityGet
     private static final Logger LOGGER = LogUtils.getLogger();
     private static final int EMPTY_TIME_NO_TICK = 300;
     private static final int MAX_SCHEDULED_TICKS_PER_TICK = 65536;
-    private final List<ServerPlayer> players = Lists.newArrayList();
+    final List<ServerPlayer> players = new java.util.concurrent.CopyOnWriteArrayList<>(); // MoraPaper - Use CopyOnWriteArrayList for safe async loops
     public final ServerChunkCache chunkSource;
     private final MinecraftServer server;
     public final io.papermc.paper.world.saveddata.PaperLevelOverrides serverLevelData; // Paper - type
@@ -604,6 +604,11 @@ public class ServerLevel extends Level implements WorldGenLevel, ServerEntityGet
     }
     // Paper end - chunk tick iteration
 
+    // MoraPaper - Get async entities from our tick list (calculated in another thread)
+    public java.util.concurrent.ConcurrentLinkedQueue<Entity> getEntitiesAsync() {
+        return entityTickList.getEntitiesAsync();
+    }
+
     public ServerLevel(
         final MinecraftServer server,
         final Executor executor,
@@ -738,6 +743,8 @@ public class ServerLevel extends Level implements WorldGenLevel, ServerEntityGet
         this.chunkDataController = new ca.spottedleaf.moonrise.patches.chunk_system.io.datacontroller.ChunkDataController((ServerLevel)(Object)this, this.chunkTaskScheduler);
         // Paper end - rewrite chunk system
         this.getCraftServer().addWorld(this.getWorld()); // CraftBukkit
+        // MoraPaper
+        net.minemora.MoraPaper.startNearestPlayerUpdater(this);
     }
 
     // Paper start
@@ -1528,6 +1535,13 @@ public class ServerLevel extends Level implements WorldGenLevel, ServerEntityGet
         });
     }
 
+    // MoraPaper - start
+    public void disableDragon() {
+        this.dragonFight = null;
+        this.paperConfig().entities.spawning.scanForLegacyEnderDragon = false;
+    }
+    // MoraPaper - end
+
     public List<? extends EnderDragon> getDragons() {
         return this.getEntities(EntityType.ENDER_DRAGON, LivingEntity::isAlive);
     }
diff --git a/net/minecraft/server/level/ServerPlayer.java b/net/minecraft/server/level/ServerPlayer.java
index 60ff8173750b57aa05bb19a3dd1c744f870f59a5..c54e42c4b733ea47fcc94b8a7cde2fbd18bf1d5e 100644
--- a/net/minecraft/server/level/ServerPlayer.java
+++ b/net/minecraft/server/level/ServerPlayer.java
@@ -231,6 +231,7 @@ public class ServerPlayer extends Player implements ca.spottedleaf.moonrise.patc
     public final ServerPlayerGameMode gameMode;
     private final PlayerAdvancements advancements;
     private final ServerStatsCounter stats;
+    public int fakePing;
     private float lastRecordedHealthAndAbsorption = Float.MIN_VALUE;
     private int lastRecordedFoodLevel = Integer.MIN_VALUE;
     private int lastRecordedAirLevel = Integer.MIN_VALUE;
@@ -485,6 +486,23 @@ public class ServerPlayer extends Player implements ca.spottedleaf.moonrise.patc
         // CraftBukkit end
     }
 
+    // MoraPaper - optimized constructor for NPCs
+    public ServerPlayer(MinecraftServer server, ServerLevel level, GameProfile gameProfile) {
+        super(level, gameProfile);
+        this.textFilter = server.createTextFilterForPlayer(this);
+        this.gameMode = server.createGameModeForPlayer(this);
+        this.server = server;
+        this.object = null;
+        // CraftBukkit start
+        this.displayName = this.getScoreboardName();
+        this.adventure$displayName = net.kyori.adventure.text.Component.text(this.getScoreboardName()); // Paper
+        this.bukkitPickUpLoot = true;
+        this.maxHealthCache = this.getMaxHealth();
+        this.recipeBook = null;
+        this.stats = null;
+        this.advancements = null;
+    }
+
     @Override
     public BlockPos adjustSpawnLocation(final ServerLevel level, final BlockPos spawnSuggestion) {
         CompletableFuture<Vec3> future = PlayerSpawnFinder.findSpawn(level, spawnSuggestion);
@@ -1529,7 +1547,7 @@ public class ServerPlayer extends Player implements ca.spottedleaf.moonrise.patc
         }
 
         if (newLevel.dimension() == lastDimension) {
-            this.connection.internalTeleport(PositionMoveRotation.of(transition), transition.relatives()); // CraftBukkit  
+            this.connection.internalTeleport(PositionMoveRotation.of(transition), transition.relatives()); // CraftBukkit
             this.connection.resetPosition();
             transition.postTeleportTransition().onTransition(this);
             return this;
@@ -2391,7 +2409,12 @@ public class ServerPlayer extends Player implements ca.spottedleaf.moonrise.patc
         this.sendOverlayMessage(Component.translatable("build.spawn_protection", pos.toShortString()).withStyle(ChatFormatting.RED));
     }
 
+    // MoraPaper - Completely disable chat
+    public boolean chatDisabled = false;
+
     public void sendSystemMessage(final Component message, final boolean overlay) {
+        if (chatDisabled) return;
+
         if (this.acceptsSystemMessages(overlay)) {
             this.connection
                 .send(
@@ -2419,6 +2442,9 @@ public class ServerPlayer extends Player implements ca.spottedleaf.moonrise.patc
         this.sendChatMessage(message, filtered, chatType, null);
     }
     public void sendChatMessage(final OutgoingChatMessage message, final boolean filtered, final ChatType.Bound chatType, @Nullable Component unsigned) {
+        // MoraPaper
+        if (chatDisabled) return;
+
         // Paper end
         if (this.acceptsChatMessages()) {
             message.sendToPlayer(this, filtered, chatType, unsigned); // Paper
diff --git a/net/minecraft/server/network/ServerGamePacketListenerImpl.java b/net/minecraft/server/network/ServerGamePacketListenerImpl.java
index 810b352503cdf56091ec6340ad72edde238fe359..e356545607250556822986bd76afd149cca53bce 100644
--- a/net/minecraft/server/network/ServerGamePacketListenerImpl.java
+++ b/net/minecraft/server/network/ServerGamePacketListenerImpl.java
@@ -1893,11 +1893,28 @@ public class ServerGamePacketListenerImpl
         }
 
         // Paper start - Teleport API
-        final Set<io.papermc.paper.entity.TeleportFlag.Relative> relativeFlags = java.util.EnumSet.noneOf(io.papermc.paper.entity.TeleportFlag.Relative.class);
+        // MoraPaper Starts - Pre teleport event
+        Set<io.papermc.paper.entity.TeleportFlag.Relative> relativeFlags = java.util.EnumSet.noneOf(io.papermc.paper.entity.TeleportFlag.Relative.class); // MoraPaper - no final
         for (final Relative relativeArgument : relatives) {
             final io.papermc.paper.entity.TeleportFlag.Relative flag = org.bukkit.craftbukkit.entity.CraftEntity.deltaRelativeToAPI(relativeArgument);
             if (flag != null) relativeFlags.add(flag);
         }
+
+        final org.bukkit.event.player.PlayerPreTeleportEvent preTeleportEvent = new org.bukkit.event.player.PlayerPreTeleportEvent(player, from, to, cause);
+        preTeleportEvent.callEvent();
+
+        if (!preTeleportEvent.getFlags().isEmpty()) {
+            // If allFlags is empty, it means it's an immutable Set, replace it with a mutable one
+            if (relativeFlags.isEmpty()) relativeFlags = new java.util.HashSet<>();
+            // Convert any TeleportFlag values to the Relative variant before adding
+            for (final io.papermc.paper.entity.TeleportFlag flag : preTeleportEvent.getFlags()) {
+                if (flag instanceof io.papermc.paper.entity.TeleportFlag.Relative rel) {
+                    relativeFlags.add(rel);
+                }
+            }
+        }
+        // MoraPaper Ends
+
         PlayerTeleportEvent event = new PlayerTeleportEvent(player, from.clone(), to.clone(), cause, java.util.Set.copyOf(relativeFlags));
         // Paper end - Teleport API
         this.cserver.getPluginManager().callEvent(event);
@@ -2570,8 +2587,9 @@ public class ServerGamePacketListenerImpl
                 LastSeenMessages result = this.lastSeenMessages.applyUpdate(update);
                 var10000 = Optional.of(result);
             } catch (LastSeenMessagesValidator.ValidationException e) {
-                LOGGER.error("Failed to validate message acknowledgements from {}: {}", this.player.getPlainTextName(), e.getMessage());
-                this.disconnectAsync(CHAT_VALIDATION_FAILED, org.bukkit.event.player.PlayerKickEvent.Cause.CHAT_VALIDATION_FAILED); // Paper - kick event causes & add proper async disconnect
+                // MoraPaper - Ignore kicks for invalid messages
+                // LOGGER.error("Failed to validate message acknowledgements from {}: {}", this.player.getPlainTextName(), e.getMessage());
+                // this.disconnectAsync(CHAT_VALIDATION_FAILED, org.bukkit.event.player.PlayerKickEvent.Cause.CHAT_VALIDATION_FAILED); // Paper - kick event causes & add proper async disconnect
                 return Optional.empty();
             }
 
@@ -2673,8 +2691,9 @@ public class ServerGamePacketListenerImpl
             try {
                 this.lastSeenMessages.applyOffset(packet.offset());
             } catch (LastSeenMessagesValidator.ValidationException e) {
-                LOGGER.error("Failed to validate message acknowledgement offset from {}: {}", this.player.getPlainTextName(), e.getMessage());
-                this.disconnectAsync(ServerGamePacketListenerImpl.CHAT_VALIDATION_FAILED, org.bukkit.event.player.PlayerKickEvent.Cause.CHAT_VALIDATION_FAILED); // Paper - kick event causes & add proper async disconnect
+                // MoraPaper - Ignore kicks for invalid messages
+                // LOGGER.error("Failed to validate message acknowledgement offset from {}: {}", this.player.getPlainTextName(), e.getMessage());
+                // this.disconnectAsync(ServerGamePacketListenerImpl.CHAT_VALIDATION_FAILED, org.bukkit.event.player.PlayerKickEvent.Cause.CHAT_VALIDATION_FAILED); // Paper - kick event causes & add proper async disconnect
             }
         }
     }
@@ -2782,6 +2801,10 @@ public class ServerGamePacketListenerImpl
             return;
         }
         // CraftBukkit end
+
+        // MoraPaper
+        if (player.chatDisabled) return;
+
         // Paper start - Ensure that client receives chat packets in the same order that we add into the message signature cache
         synchronized (this.messageSignatureCache) {
         this.send(
diff --git a/net/minecraft/world/entity/LivingEntity.java b/net/minecraft/world/entity/LivingEntity.java
index d10d082046e41cf937f1a439e3e7dd61cd16f326..ea489dd609129d1722cc12d9a9cc32f0a15f7d8e 100644
--- a/net/minecraft/world/entity/LivingEntity.java
+++ b/net/minecraft/world/entity/LivingEntity.java
@@ -4903,4 +4903,9 @@ public abstract class LivingEntity extends Entity implements Attackable, Waypoin
 
     public record Fallsounds(SoundEvent small, SoundEvent big) {
     }
+
+    // MoraPaper start - LivingEntity equipment accessor
+    public EntityEquipment getEquipment() {
+        return this.equipment;
+    }
 }
diff --git a/net/minecraft/world/entity/Mob.java b/net/minecraft/world/entity/Mob.java
index 8b036640be655c65852e0862e30962e443d3024d..9bcacbbe7afe6e335af5615f3f8f507591b1efaa 100644
--- a/net/minecraft/world/entity/Mob.java
+++ b/net/minecraft/world/entity/Mob.java
@@ -150,6 +150,7 @@ public abstract class Mob extends LivingEntity implements Targeting, EquipmentUs
     private int homeRadius = -1;
     public boolean aware = true; // CraftBukkit
     public net.kyori.adventure.util.TriState despawnInPeacefulOverride = net.kyori.adventure.util.TriState.NOT_SET; // Paper - allow changing despawnInPeaceful
+    public boolean ignoreDisableAI = false;
 
     protected Mob(final EntityType<? extends Mob> type, final Level level) {
         super(type, level);
@@ -529,8 +530,13 @@ public abstract class Mob extends LivingEntity implements Targeting, EquipmentUs
         this.resetAngularLeashMomentum();
     }
 
+    // MoraPaper
+    public static int distanceSqrToDisableAi = 400;
+
     @Override
     public void aiStep() {
+        if (!ignoreDisableAI && (nearestPlayerData.distanceSqr > distanceSqrToDisableAi)) return;
+
         super.aiStep();
         if (this.is(EntityTypeTags.BURN_IN_DAYLIGHT)) {
             this.burnUndead();
@@ -758,21 +764,39 @@ public abstract class Mob extends LivingEntity implements Targeting, EquipmentUs
     }
     // Paper end - allow changing despawnInPeaceful
 
+    // MoraPaper
+    public net.minemora.MoraPaper.PlayerDistanceData nearestPlayerData = new net.minemora.MoraPaper.PlayerDistanceData();
+
+    public @Nullable Player getNearestPlayerNew() {
+        if(nearestPlayerData.uuid != null) {
+            org.bukkit.entity.Player bukkitPlayer = org.bukkit.Bukkit.getPlayer(nearestPlayerData.uuid);
+            if(bukkitPlayer != null) {
+                return ((org.bukkit.craftbukkit.entity.CraftPlayer)bukkitPlayer).getHandle();
+            }
+            else {
+                nearestPlayerData.uuid = null;
+            }
+        }
+        return null;
+    }
+
     @Override
     public void checkDespawn() {
         if (this.level().getDifficulty() == Difficulty.PEACEFUL && this.shouldDespawnInPeaceful()) { // Paper - allow changing despawnInPeaceful
             this.discard(EntityRemoveEvent.Cause.DESPAWN); // CraftBukkit - add Bukkit remove cause
         } else if (!this.isPersistenceRequired() && !this.requiresCustomPersistence()) {
-            Entity player = this.level().findNearbyPlayer(this, -1.0, EntitySelector.PLAYER_AFFECTS_SPAWNING); // Paper - Affects Spawning API
+            Player player = this.getNearestPlayerNew(); // MoraPaper - Optimize get nearest player
             if (player != null) {
                 // Paper start - Configurable despawn distances
+                // MoraPaper Start - Use our cached nearest player distance
                 final io.papermc.paper.configuration.WorldConfiguration.Entities.Spawning.DespawnRangePair despawnRangePair = this.level().paperConfig().entities.spawning.despawnRanges.get(this.getType().getCategory());
                 final io.papermc.paper.configuration.type.DespawnRange.Shape shape = this.level().paperConfig().entities.spawning.despawnRangeShape;
-                final double dy = Math.abs(player.getY() - this.getY());
-                final double dySqr = Mth.square(dy);
-                final double dxSqr = Mth.square(player.getX() - this.getX());
-                final double dzSqr = Mth.square(player.getZ() - this.getZ());
-                final double distanceSquared = dxSqr + dzSqr + dySqr;
+                final double dy = nearestPlayerData.dy;
+                final double dySqr = nearestPlayerData.dySqr;
+                final double dxSqr = nearestPlayerData.dxSqr;
+                final double dzSqr = nearestPlayerData.dzSqr;
+                final double distanceSquared = nearestPlayerData.distanceSqr;
+                // MoraPaper End
                 // Despawn if hard/soft limit is exceeded
                 if (despawnRangePair.hard().shouldDespawn(shape, dxSqr, dySqr, dzSqr, dy) && this.removeWhenFarAway(distanceSquared)) {
                     this.discard(EntityRemoveEvent.Cause.DESPAWN); // CraftBukkit - add Bukkit remove cause
diff --git a/net/minecraft/world/entity/ai/goal/GoalSelector.java b/net/minecraft/world/entity/ai/goal/GoalSelector.java
index 1ddf8796d14d472bec48546f25ac4583d483c8e9..7e8a66870aeb70e8fab76dd7c1fcf18bcdef45aa 100644
--- a/net/minecraft/world/entity/ai/goal/GoalSelector.java
+++ b/net/minecraft/world/entity/ai/goal/GoalSelector.java
@@ -38,7 +38,7 @@ public class GoalSelector {
     // Paper start - EAR 2
     public boolean inactiveTick() {
         this.curRate++;
-        return this.curRate % 3 == 0; // TODO newGoalRate was already unused in 1.20.4, check if this is correct
+        return this.curRate % 40 == 0; // MoraPaper - Increase curRate from 3 to 40
     }
 
     public boolean hasTasks() {
diff --git a/net/minecraft/world/entity/boss/enderdragon/EnderDragon.java b/net/minecraft/world/entity/boss/enderdragon/EnderDragon.java
index 6cccc63f742e25de2655cf092a5cd2b861b59978..f279bc50fd8bdaaa60e1cef7ffdb604c8c79f5ce 100644
--- a/net/minecraft/world/entity/boss/enderdragon/EnderDragon.java
+++ b/net/minecraft/world/entity/boss/enderdragon/EnderDragon.java
@@ -672,6 +672,12 @@ public class EnderDragon extends Mob implements Enemy {
         }
     }
 
+    // MoraPaper - Add the offset to the vector based on origin
+    public void addOrigin(Vec3 vec3) {
+        final net.minecraft.core.BlockPos fightOrigin = getFightOrigin();
+        vec3.add(fightOrigin.getX(), 0, fightOrigin.getZ());
+    }
+
     public int findClosestNode() {
         if (this.nodes[0] == null) {
             for (int i = 0; i < 24; i++) {
@@ -693,6 +699,10 @@ public class EnderDragon extends Mob implements Enemy {
                     nodeZ = Mth.floor(20.0F * Mth.sin(2.0F * (-Mth.PI + (float) (Math.PI / 4) * multiplier)));
                 }
 
+                // MoraPaper - Offset nodes to origin
+                nodeX += getFightOrigin().getX();
+                nodeZ += getFightOrigin().getZ();
+
                 int nodeY = Math.max(
                     73, this.level().getHeightmapPos(Heightmap.Types.MOTION_BLOCKING_NO_LEAVES, new BlockPos(nodeX, 0, nodeZ)).getY() + yAdjustment
                 );
@@ -900,7 +910,7 @@ public class EnderDragon extends Mob implements Enemy {
         EnderDragonPhase<? extends DragonPhaseInstance> phase = phaseInstance.getPhase();
         Vec3 result;
         if (phase == EnderDragonPhase.LANDING || phase == EnderDragonPhase.TAKEOFF) {
-            BlockPos egg = this.level().getHeightmapPos(Heightmap.Types.MOTION_BLOCKING_NO_LEAVES, this.getPodium()); // Paper - Allow changing the EnderDragon podium
+            BlockPos egg = calculateLandPosition((net.minecraft.server.level.ServerLevel) this.level(), net.minecraft.world.level.levelgen.Heightmap.Types.MOTION_BLOCKING_NO_LEAVES); // MoraPaper - Allow changing the EnderDragon podium
             float dist = Math.max((float)Math.sqrt(egg.distToCenterSqr(this.position())) / 4.0F, 1.0F);
             float yOffset = 6.0F / dist;
             float xRotOld = this.getXRot();
@@ -1001,5 +1011,11 @@ public class EnderDragon extends Mob implements Enemy {
 
         return shouldDrop ? xpCount : 0;
     }
+
+    // MoraSpigot - Allow overriding of land position to ignore the heightmap position
+    public net.minecraft.core.BlockPos calculateLandPosition(final net.minecraft.server.level.ServerLevel level, final net.minecraft.world.level.levelgen.Heightmap.Types heightmapType) {
+        return level.getHeightmapPos(heightmapType, getPodium()); // Paper - Allow changing the EnderDragon podium
+    }
+
     // CraftBukkit end
 }
diff --git a/net/minecraft/world/entity/boss/enderdragon/phases/DragonDeathPhase.java b/net/minecraft/world/entity/boss/enderdragon/phases/DragonDeathPhase.java
index f2555c5ce49a6e6b5c9eea2b9d2ff1d7d65bef6c..602ed5153f6be387cac2aeb7e5534847cadbba9c 100644
--- a/net/minecraft/world/entity/boss/enderdragon/phases/DragonDeathPhase.java
+++ b/net/minecraft/world/entity/boss/enderdragon/phases/DragonDeathPhase.java
@@ -5,7 +5,6 @@ import net.minecraft.core.particles.ParticleTypes;
 import net.minecraft.server.level.ServerLevel;
 import net.minecraft.world.entity.boss.enderdragon.EnderDragon;
 import net.minecraft.world.level.levelgen.Heightmap;
-import net.minecraft.world.level.levelgen.feature.EndPodiumFeature;
 import net.minecraft.world.phys.Vec3;
 import org.jspecify.annotations.Nullable;
 
@@ -33,7 +32,7 @@ public class DragonDeathPhase extends AbstractDragonPhaseInstance {
     public void doServerTick(final ServerLevel level) {
         this.time++;
         if (this.targetLocation == null) {
-            BlockPos egg = level.getHeightmapPos(Heightmap.Types.MOTION_BLOCKING, this.dragon.getPodium()); // Paper - Allow changing the EnderDragon podium
+            BlockPos egg = this.dragon.calculateLandPosition(level, Heightmap.Types.MOTION_BLOCKING); // MoraPaper - Allow changing the EnderDragon podium
             this.targetLocation = Vec3.atBottomCenterOf(egg);
         }
 
diff --git a/net/minecraft/world/entity/boss/enderdragon/phases/DragonHoldingPatternPhase.java b/net/minecraft/world/entity/boss/enderdragon/phases/DragonHoldingPatternPhase.java
index 5245e9a1c4eb289efcf832cea4aebee1dcf02daf..d28e224fb67825b08291ebc46904351d3899670f 100644
--- a/net/minecraft/world/entity/boss/enderdragon/phases/DragonHoldingPatternPhase.java
+++ b/net/minecraft/world/entity/boss/enderdragon/phases/DragonHoldingPatternPhase.java
@@ -50,7 +50,7 @@ public class DragonHoldingPatternPhase extends AbstractDragonPhaseInstance {
 
     private void findNewTarget(final ServerLevel level) {
         if (this.currentPath != null && this.currentPath.isDone()) {
-            BlockPos egg = level.getHeightmapPos(Heightmap.Types.MOTION_BLOCKING_NO_LEAVES, this.dragon.getPodium()); // Paper - Allow changing the EnderDragon podium
+            BlockPos egg = this.dragon.calculateLandPosition(level, Heightmap.Types.MOTION_BLOCKING_NO_LEAVES); // MoraPaper - Allow changing the EnderDragon podium
             int crystals = this.dragon.getDragonFight() == null ? 0 : this.dragon.getDragonFight().aliveCrystals();
             if (this.dragon.getRandom().nextInt(crystals + 3) == 0) {
                 this.dragon.getPhaseManager().setPhase(EnderDragonPhase.LANDING_APPROACH);
diff --git a/net/minecraft/world/entity/boss/enderdragon/phases/DragonLandingApproachPhase.java b/net/minecraft/world/entity/boss/enderdragon/phases/DragonLandingApproachPhase.java
index 304356cda2d188bbfda8a6be27b08f1b8685e7dc..31d46296a28d857a53011e27f2ffcc230713e07f 100644
--- a/net/minecraft/world/entity/boss/enderdragon/phases/DragonLandingApproachPhase.java
+++ b/net/minecraft/world/entity/boss/enderdragon/phases/DragonLandingApproachPhase.java
@@ -6,8 +6,6 @@ import net.minecraft.server.level.ServerLevel;
 import net.minecraft.world.entity.ai.targeting.TargetingConditions;
 import net.minecraft.world.entity.boss.enderdragon.EnderDragon;
 import net.minecraft.world.entity.player.Player;
-import net.minecraft.world.level.levelgen.Heightmap;
-import net.minecraft.world.level.levelgen.feature.EndPodiumFeature;
 import net.minecraft.world.level.pathfinder.Node;
 import net.minecraft.world.level.pathfinder.Path;
 import net.minecraft.world.phys.Vec3;
@@ -49,14 +47,22 @@ public class DragonLandingApproachPhase extends AbstractDragonPhaseInstance {
     private void findNewTarget(final ServerLevel level) {
         if (this.currentPath == null || this.currentPath.isDone()) {
             int currentNodeIndex = this.dragon.findClosestNode();
-            BlockPos egg = level.getHeightmapPos(Heightmap.Types.MOTION_BLOCKING_NO_LEAVES, this.dragon.getPodium()); // Paper - Allow changing the EnderDragon podium
+            BlockPos egg = this.dragon.calculateLandPosition(level, net.minecraft.world.level.levelgen.Heightmap.Types.MOTION_BLOCKING_NO_LEAVES); // MoraPaper - Allow changing the EnderDragon podium
             Player playerNearestToEgg = level.getNearestPlayer(NEAR_EGG_TARGETING, this.dragon, egg.getX(), egg.getY(), egg.getZ());
             int targetNodeIndex;
             if (playerNearestToEgg != null) {
                 Vec3 aim = new Vec3(playerNearestToEgg.getX(), 0.0, playerNearestToEgg.getZ()).normalize();
-                targetNodeIndex = this.dragon.findClosestNode(-aim.x * 40.0, 105.0, -aim.z * 40.0);
+                // MoraPaper - Add origin
+                Vec3 extra = new Vec3(-aim.x * 40.0, 105.0, -aim.z * 40.0);
+                this.dragon.addOrigin(extra);
+
+                targetNodeIndex = this.dragon.findClosestNode(extra.x, extra.y, -extra.z);
             } else {
-                targetNodeIndex = this.dragon.findClosestNode(40.0, egg.getY(), 0.0);
+                // MoraPaper - Add origin
+                Vec3 extra = new Vec3(40.0, egg.getY(), 0.0);
+                this.dragon.addOrigin(extra);
+
+                targetNodeIndex = this.dragon.findClosestNode(extra.x, extra.y, extra.z);
             }
 
             Node finalNode = new Node(egg.getX(), egg.getY(), egg.getZ());
diff --git a/net/minecraft/world/entity/boss/enderdragon/phases/DragonLandingPhase.java b/net/minecraft/world/entity/boss/enderdragon/phases/DragonLandingPhase.java
index e5089f65bc6bb4df4fa7800399642730e04bad66..4787114aebf7d9dce891bd3701e3579ebc96f4fa 100644
--- a/net/minecraft/world/entity/boss/enderdragon/phases/DragonLandingPhase.java
+++ b/net/minecraft/world/entity/boss/enderdragon/phases/DragonLandingPhase.java
@@ -50,7 +50,7 @@ public class DragonLandingPhase extends AbstractDragonPhaseInstance {
     public void doServerTick(final ServerLevel level) {
         if (this.targetLocation == null) {
             this.targetLocation = Vec3.atBottomCenterOf(
-                level.getHeightmapPos(Heightmap.Types.MOTION_BLOCKING_NO_LEAVES, this.dragon.getPodium()) // Paper - Allow changing the EnderDragon podium
+                this.dragon.calculateLandPosition(level, net.minecraft.world.level.levelgen.Heightmap.Types.MOTION_BLOCKING_NO_LEAVES)  // MoraPaper - Allow changing the EnderDragon podium
             );
         }
 
diff --git a/net/minecraft/world/entity/boss/enderdragon/phases/DragonTakeoffPhase.java b/net/minecraft/world/entity/boss/enderdragon/phases/DragonTakeoffPhase.java
index 64ee967af3ab2b934660ca521548826151c6c08e..8dc0ff1281d41911edc79f6df6b1208dd8da071f 100644
--- a/net/minecraft/world/entity/boss/enderdragon/phases/DragonTakeoffPhase.java
+++ b/net/minecraft/world/entity/boss/enderdragon/phases/DragonTakeoffPhase.java
@@ -22,7 +22,7 @@ public class DragonTakeoffPhase extends AbstractDragonPhaseInstance {
     @Override
     public void doServerTick(final ServerLevel level) {
         if (!this.firstTick && this.currentPath != null) {
-            BlockPos egg = level.getHeightmapPos(Heightmap.Types.MOTION_BLOCKING_NO_LEAVES, this.dragon.getPodium()); // Paper - Allow changing the EnderDragon podium
+            BlockPos egg = this.dragon.calculateLandPosition(level, Heightmap.Types.MOTION_BLOCKING_NO_LEAVES); // MoraPaper - Allow changing the EnderDragon podium
             if (!egg.closerToCenterThan(this.dragon.position(), 10.0)) {
                 this.dragon.getPhaseManager().setPhase(EnderDragonPhase.HOLDING_PATTERN);
             }
@@ -42,7 +42,12 @@ public class DragonTakeoffPhase extends AbstractDragonPhaseInstance {
     private void findNewTarget() {
         int currentNodeIndex = this.dragon.findClosestNode();
         Vec3 lookVector = this.dragon.getHeadLookVector(1.0F);
-        int targetNodeIndex = this.dragon.findClosestNode(-lookVector.x * 40.0, 105.0, -lookVector.z * 40.0);
+
+        // MoraPaper - Add origin
+        Vec3 extra = new Vec3(-lookVector.x * 40.0, 105.0, -lookVector.z * 40.0);
+        this.dragon.addOrigin(extra);
+
+        int targetNodeIndex = this.dragon.findClosestNode(extra.x, extra.y, extra.z);
         if (this.dragon.getDragonFight() != null && this.dragon.getDragonFight().aliveCrystals() > 0) {
             targetNodeIndex %= 12;
             if (targetNodeIndex < 0) {
diff --git a/net/minecraft/world/entity/monster/Monster.java b/net/minecraft/world/entity/monster/Monster.java
index c55b51ed7ed27cc658170d5c35c81a818b8867f2..b74573d0ac8e93bbdbd93ba2b7e5568337d246fc 100644
--- a/net/minecraft/world/entity/monster/Monster.java
+++ b/net/minecraft/world/entity/monster/Monster.java
@@ -41,6 +41,7 @@ public abstract class Monster extends PathfinderMob implements Enemy {
 
     @Override
     public void aiStep() {
+        if (!ignoreDisableAI && (nearestPlayerData.distanceSqr > distanceSqrToDisableAi)) return;
         this.updateSwingTime();
         this.updateNoActionTime();
         super.aiStep();
diff --git a/net/minecraft/world/entity/npc/villager/Villager.java b/net/minecraft/world/entity/npc/villager/Villager.java
index e4ba6e200e85f4e226d6a3e68db734f277b89226..4c5c5b56e9f86c121018ee7eb6fb163144ce47cc 100644
--- a/net/minecraft/world/entity/npc/villager/Villager.java
+++ b/net/minecraft/world/entity/npc/villager/Villager.java
@@ -268,7 +268,12 @@ public class Villager extends AbstractVillager implements VillagerDataHolder, Re
         // Paper start - EAR 2
         this.customServerAiStep(level, false);
     }
-    protected void customServerAiStep(ServerLevel level, final boolean inactive) {
+    protected void customServerAiStep(ServerLevel level, boolean inactive) {
+        // MoraSpigot - Mark inactive if the player is too far away
+        if(nearestPlayerData.distanceSqr > distanceSqrToDisableAi) {
+            inactive = true;
+        }
+
         // Paper end - EAR 2
         ProfilerFiller profiler = Profiler.get();
         profiler.push("villagerBrain");
diff --git a/net/minecraft/world/entity/projectile/throwableitemprojectile/Snowball.java b/net/minecraft/world/entity/projectile/throwableitemprojectile/Snowball.java
index 6e828818909df2b31349398639257b7a6f89fcd5..084d4c3e08df5147484c89d946db72fea3dd11de 100644
--- a/net/minecraft/world/entity/projectile/throwableitemprojectile/Snowball.java
+++ b/net/minecraft/world/entity/projectile/throwableitemprojectile/Snowball.java
@@ -66,4 +66,13 @@ public class Snowball extends ThrowableItemProjectile {
             this.discard(org.bukkit.event.entity.EntityRemoveEvent.Cause.HIT); // CraftBukkit - add Bukkit remove cause
         }
     }
+
+    // MoraPaper start - Prevent Snowball Lag Machine
+    @Override
+    public void tick() {
+        super.tick();
+        if (this.tickCount >= 8) discard();
+    }
+    // MoraPaper end - Prevent Snowball Lag Machine
+
 }
diff --git a/net/minecraft/world/entity/projectile/throwableitemprojectile/ThrownEgg.java b/net/minecraft/world/entity/projectile/throwableitemprojectile/ThrownEgg.java
index 55d97e1a69d17308196938398ab5a34bcca44e46..af9bc3b89d7561de3c1a6e9ab011f28345e27fbd 100644
--- a/net/minecraft/world/entity/projectile/throwableitemprojectile/ThrownEgg.java
+++ b/net/minecraft/world/entity/projectile/throwableitemprojectile/ThrownEgg.java
@@ -131,4 +131,13 @@ public class ThrownEgg extends ThrowableItemProjectile {
     public Item getDefaultItem() {
         return Items.EGG;
     }
+
+    // MoraPaper start - Prevent Egg Lag Machine
+    @Override
+    public void tick() {
+        super.tick();
+        if (this.tickCount >= 8) discard();
+    }
+    // MoraPaper end - Prevent Egg Lag Machine
+
 }
diff --git a/net/minecraft/world/entity/vehicle/minecart/NewMinecartBehavior.java b/net/minecraft/world/entity/vehicle/minecart/NewMinecartBehavior.java
index 18f3ff2cb03cc401c999acdf31bb1812e69e0f62..15464dc3fd102b74bd2362278adf161a5f6aa73c 100644
--- a/net/minecraft/world/entity/vehicle/minecart/NewMinecartBehavior.java
+++ b/net/minecraft/world/entity/vehicle/minecart/NewMinecartBehavior.java
@@ -368,6 +368,13 @@ public class NewMinecartBehavior extends MinecartBehavior {
 
     private Vec3 calculatePlayerInputSpeed(final Vec3 deltaMovement) {
         if (this.minecart.getFirstPassenger() instanceof ServerPlayer player) {
+            // MoraPaper Start - Steer minecart event
+            final org.bukkit.event.player.PlayerSteerMinecartEvent event = new org.bukkit.event.player.PlayerSteerMinecartEvent((org.bukkit.entity.Player) player.getBukkitEntity(), (org.bukkit.entity.Minecart) this.minecart.getBukkitEntity());
+            event.callEvent();
+
+            if (event.isCancelled()) return deltaMovement;
+            // MoraPaper End
+
             Vec3 moveIntent = player.getLastClientMoveIntent();
             if (moveIntent.lengthSqr() > 0.0) {
                 Vec3 riderMovement = moveIntent.normalize();
diff --git a/net/minecraft/world/entity/vehicle/minecart/OldMinecartBehavior.java b/net/minecraft/world/entity/vehicle/minecart/OldMinecartBehavior.java
index 3c8694e75cabd68690d9139ecedc9e079db482c5..c57fd515986e60114baa205e33d6ada88a450b7d 100644
--- a/net/minecraft/world/entity/vehicle/minecart/OldMinecartBehavior.java
+++ b/net/minecraft/world/entity/vehicle/minecart/OldMinecartBehavior.java
@@ -162,6 +162,16 @@ public class OldMinecartBehavior extends MinecartBehavior {
         Vec3 moveIntent;
         if (this.minecart.getFirstPassenger() instanceof ServerPlayer player) {
             moveIntent = player.getLastClientMoveIntent();
+            // MoraPaper Start - Steer minecart event
+            if (moveIntent.x != 0 || moveIntent.y != 0 || moveIntent.z != 0) {
+                final org.bukkit.event.player.PlayerSteerMinecartEvent event = new org.bukkit.event.player.PlayerSteerMinecartEvent((org.bukkit.entity.Player) player.getBukkitEntity(), (org.bukkit.entity.Minecart) this.minecart.getBukkitEntity());
+                event.callEvent();
+
+                if (event.isCancelled()) {
+                    moveIntent = net.minecraft.world.phys.Vec3.ZERO;
+                }
+            }
+            // MoraPaper End
         } else {
             moveIntent = Vec3.ZERO;
         }
diff --git a/net/minecraft/world/level/entity/EntityTickList.java b/net/minecraft/world/level/entity/EntityTickList.java
index a30409eddfaded878a3ac54bb526fd2e4b887c57..745b02356a9db225334adf200cb19ed18760aa9d 100644
--- a/net/minecraft/world/level/entity/EntityTickList.java
+++ b/net/minecraft/world/level/entity/EntityTickList.java
@@ -11,6 +11,9 @@ import org.jspecify.annotations.Nullable;
 public class EntityTickList {
     private final ca.spottedleaf.moonrise.common.list.IteratorSafeOrderedReferenceSet<net.minecraft.world.entity.Entity> entities = new ca.spottedleaf.moonrise.common.list.IteratorSafeOrderedReferenceSet<>(); // Paper - rewrite chunk system
 
+    // MoraPaper - async list to use entities from other threads
+    private final java.util.concurrent.ConcurrentLinkedQueue<net.minecraft.world.entity.Entity> asyncEntities = new java.util.concurrent.ConcurrentLinkedQueue<>();
+
     private void ensureActiveIsNotIterated() {
         // Paper - rewrite chunk system
     }
@@ -18,11 +21,13 @@ public class EntityTickList {
     public void add(final Entity entity) {
         this.ensureActiveIsNotIterated();
         this.entities.add(entity); // Paper - rewrite chunk system
+        this.asyncEntities.add(entity); // MoraPaper - async list to use entities from other threads
     }
 
     public void remove(final Entity entity) {
         this.ensureActiveIsNotIterated();
         this.entities.remove(entity); // Paper - rewrite chunk system
+        this.asyncEntities.remove(entity); // MoraPaper - async list to use entities from other threads
     }
 
     public boolean contains(final Entity entity) {
@@ -43,4 +48,10 @@ public class EntityTickList {
         }
         // Paper end - rewrite chunk system
     }
+
+    // MoraPaper - async list to use entities from other threads
+    public java.util.concurrent.ConcurrentLinkedQueue<Entity> getEntitiesAsync() {
+        return asyncEntities;
+    }
+
 }
